File: TidbitsTrivia/Core/Engine/BotOpponent.swift
import Foundation
import Observation

/// Play vs CPU — the online-multiplayer v0 (Decision 038,
/// docs/ONLINE-MULTIPLAYER-PLAYBOOK.md §5). A bot is a small parameterized
/// model resolved locally against the same question the player sees; later
/// (v1) the identical resolution runs inside the server room actor as the
/// timeout-fill / dropout brain.
///
/// HONESTY RULE (learning-orientation, non-negotiable): a bot is always
/// visibly labeled CPU. Never present a bot as a human.
///
/// Believability levers (the playbook's spec): non-extreme, category-varying
/// correct-rates (visible strengths/weaknesses) + jittered right-skewed
/// timing with an occasional freeze — a bot that answers in exactly 3.0s
/// every time is the tell. Mirrors: `Bots.kt` (Android) and `js/bots.js`
/// (web); keep the three in lockstep.
nonisolated struct BotProfile: Identifiable, Hashable, Sendable {
    let id: String
    let name: String
    /// Global correct-rate at medium difficulty (0…1, never extreme).
    let baseSkill: Double
    /// Per-category offset — strengths/weaknesses (e.g. Sports +0.15).
    let categorySkill: [String: Double]
    /// Answer-time model: log-normal around `speedMean` seconds.
    let speedMean: Double
    let speedSigma: Double

    static let rookie = BotProfile(
        id: "rookie", name: "Rookie Rae", baseSkill: 0.55,
        categorySkill: ["sports": 0.15, "film": 0.10, "science": -0.12],
        speedMean: 6.5, speedSigma: 0.45)
    static let regular = BotProfile(
        id: "regular", name: "Trivia Tina", baseSkill: 0.70,
        categorySkill: ["history": 0.10, "arts": 0.08, "sports": -0.10],
        speedMean: 5.5, speedSigma: 0.40)
    static let ace = BotProfile(
        id: "ace", name: "Ace Botsworth", baseSkill: 0.85,
        categorySkill: ["science": 0.10, "geography": 0.08, "music": -0.08],
        speedMean: 4.0, speedSigma: 0.35)

    /// The adaptive sparring partner: tracks the player's recent accuracy so
    /// solo-vs-CPU stays a fair fight ("meet the learner where they are").
    static func house(playerAccuracy: Double) -> BotProfile {
        BotProfile(
            id: "house", name: "The House", baseSkill: min(0.90, max(0.35, playerAccuracy)),
            categorySkill: [:], speedMean: 5.0, speedSigma: 0.40)
    }

    static let presets: [BotProfile] = [.rookie, .regular, .ace]
}

/// One bot's resolution for one question. `seconds == nil` = froze / ran out
/// of clock (humans do too — that variance is the believability payload).
nonisolated struct BotAnswer: Sendable {
    let botID: String
    let choiceIndex: Int?
    let seconds: Double?
    var answered: Bool { choiceIndex != nil }
}

nonisolated enum BotBrain {
    static func difficultyAdj(_ difficulty: Int) -> Double {
        if difficulty <= 2 { return 0.15 }
        if difficulty >= 4 { return -0.20 }
        return 0
    }

    /// Resolve what this bot does with this question, inside `window` seconds.
    static func resolve(_ bot: BotProfile, categoryID: String, difficulty: Int,
                        correctIndex: Int, optionCount: Int, window: Double) -> BotAnswer {
        let p = min(0.98, max(0.02,
            bot.baseSkill + (bot.categorySkill[categoryID] ?? 0) + difficultyAdj(difficulty)))
        // ~5% freeze: no answer at all.
        if Double.random(in: 0..<1) < 0.05 {
            return BotAnswer(botID: bot.id, choiceIndex: nil, seconds: nil)
        }
        let correct = Double.random(in: 0..<1) < p
        // Log-normal, right-skewed; knowing feels fast, so correct trims ~15%.
        let gauss = Self.gaussian()
        var t = exp(log(bot.speedMean) + gauss * bot.speedSigma)
        if correct { t *= 0.85 }
        t = min(max(t, 0.8), max(1.0, window - 0.5))
        let choice: Int
        if correct {
            choice = correctIndex
        } else {
            var wrong = Array(0..<max(optionCount, 2))
            wrong.removeAll { $0 == correctIndex }
            choice = wrong.randomElement() ?? 0
        }
        return BotAnswer(botID: bot.id, choiceIndex: choice, seconds: t)
    }

    private static func gaussian() -> Double {
        // Box–Muller.
        let u1 = Double.random(in: Double.ulpOfOne..<1)
        let u2 = Double.random(in: 0..<1)
        return sqrt(-2 * log(u1)) * cos(2 * .pi * u2)
    }
}

/// The running vs-CPU match: per-bot scores/streaks beside the player's own
/// engine-scored game. Resolve at question start, commit at reveal — so the
/// reveal can show what each opponent did on the SAME question.
@Observable
@MainActor
final class BotMatch {
    struct Seat: Identifiable {
        let bot: BotProfile
        var score = 0
        var streak = 0
        var lastCorrect: Bool?
        var id: String { bot.id }
    }

    private(set) var seats: [Seat]
    private(set) var pending: [BotAnswer] = []
    private var committedIndex = -1

    init(bots: [BotProfile]) {
        seats = bots.map { Seat(bot: $0) }
    }

    func beginQuestion(_ q: Question, window: Double) {
        pending = seats.map {
            BotBrain.resolve($0.bot, categoryID: q.categoryID, difficulty: q.difficulty,
                             correctIndex: q.correctIndex, optionCount: q.options.count, window: window)
        }
    }

    /// Apply the pending answers with the SAME Scoring rules the player gets.
    func commit(question q: Question, index: Int, budget: Double) {
        guard index != committedIndex else { return }   // reveal fires once
        committedIndex = index
        for answer in pending {
            guard let i = seats.firstIndex(where: { $0.bot.id == answer.botID }) else { continue }
            let correct = answer.choiceIndex == q.correctIndex
            seats[i].lastCorrect = answer.answered ? correct : false
            if correct {
                seats[i].streak += 1
                seats[i].score += Scoring.points(correct: true, secondsTaken: answer.seconds ?? budget,
                                                 budget: budget, streak: seats[i].streak)
            } else {
                seats[i].streak = 0
            }
        }
    }

    /// Standings rows (bot side only — the caller weaves in the human player).
    var standings: [Seat] { seats.sorted { $0.score > $1.score } }
}

File: windows/Tidbits.Core/Networking/BotOpponent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Tidbits.Core.Engine;
using Tidbits.Core.Models;

namespace Tidbits.Core.Networking;

/// Play vs CPU — the online-multiplayer v0 (Decision 038). Faithful C# twin of
/// js/bots.js / BotOpponent.swift / Bots.kt (keep the four in lockstep).
///
/// HONESTY RULE (learning-orientation, non-negotiable): a bot is always visibly
/// labeled CPU. Never present a bot as a human.
public sealed record Bot(string Id, string Name, double BaseSkill,
    IReadOnlyDictionary<string, double> CategorySkill, double SpeedMean, double SpeedSigma);

/// What a bot did with one question, within the answer window.
public readonly record struct BotResult(string BotId, int? ChoiceIndex, double? Seconds);

public static class Bots
{
    public static readonly IReadOnlyDictionary<string, Bot> All = new Dictionary<string, Bot>
    {
        ["rookie"] = new("rookie", "Rookie Rae", 0.55,
            new Dictionary<string, double> { ["sports"] = 0.15, ["film"] = 0.10, ["science"] = -0.12 }, 6.5, 0.45),
        ["regular"] = new("regular", "Trivia Tina", 0.70,
            new Dictionary<string, double> { ["history"] = 0.10, ["arts"] = 0.08, ["sports"] = -0.10 }, 5.5, 0.40),
        ["ace"] = new("ace", "Ace Botsworth", 0.85,
            new Dictionary<string, double> { ["science"] = 0.10, ["geography"] = 0.08, ["music"] = -0.08 }, 4.0, 0.35),
    };

    /// Adapts to the player's recent accuracy so solo-vs-CPU stays a fair fight.
    public static Bot House(double playerAccuracy) =>
        new("house", "The House", Math.Min(0.90, Math.Max(0.35, playerAccuracy)),
            new Dictionary<string, double>(), 5.0, 0.40);

    public static Bot ById(string id, double playerAccuracy) =>
        All.TryGetValue(id, out var b) ? b : House(playerAccuracy);

    private static double DifficultyAdj(int d) => d <= 2 ? 0.15 : d >= 4 ? -0.20 : 0;

    private static double Gaussian(Random rng) // Box–Muller
    {
        var u1 = Math.Max(rng.NextDouble(), double.Epsilon);
        var u2 = rng.NextDouble();
        return Math.Sqrt(-2 * Math.Log(u1)) * Math.Cos(2 * Math.PI * u2);
    }

    /// Resolve what this bot does with this question, inside `window` seconds.
    public static BotResult Resolve(Bot bot, string categoryId, int difficulty, int correctIndex,
        int optionCount, double windowSecs, Random rng)
    {
        var catAdj = bot.CategorySkill.TryGetValue(categoryId, out var c) ? c : 0;
        var p = Math.Min(0.98, Math.Max(0.02, bot.BaseSkill + catAdj + DifficultyAdj(difficulty)));
        if (rng.NextDouble() < 0.05) return new(bot.Id, null, null); // freeze
        var correct = rng.NextDouble() < p;
        var t = Math.Exp(Math.Log(bot.SpeedMean) + Gaussian(rng) * bot.SpeedSigma);
        if (correct) t *= 0.85; // knowing feels fast
        t = Math.Min(Math.Max(t, 0.8), Math.Max(1.0, windowSecs - 0.5));
        var choice = correctIndex;
        if (!correct)
        {
            var wrong = new List<int>();
            for (int i = 0; i < Math.Max(optionCount, 2); i++) if (i != correctIndex) wrong.Add(i);
            choice = wrong[rng.Next(wrong.Count)];
        }
        return new(bot.Id, choice, t);
    }
}

/// The running vs-CPU match beside the player's engine-scored game.
public sealed class VsMatch
{
    public sealed class Seat
    {
        public required Bot Bot { get; init; }
        public int Score { get; set; }
        public int Streak { get; set; }
        public bool? LastCorrect { get; set; }
    }

    private readonly Random _rng;
    private readonly List<Seat> _seats;
    private List<BotResult> _pending = new();
    private int _committed = -1;

    public VsMatch(IEnumerable<Bot> bots, Random? rng = null)
    {
        _rng = rng ?? new Random();
        _seats = bots.Select(b => new Seat { Bot = b }).ToList();
    }

    public IReadOnlyList<Seat> Seats => _seats;

    public void BeginQuestion(Question q, double windowSecs)
    {
        _pending = _seats.Select(s =>
            Bots.Resolve(s.Bot, q.CategoryId, q.Difficulty, q.CorrectIndex, q.Options.Count, windowSecs, _rng)).ToList();
    }

    public void Commit(Question q, int index, double budget)
    {
        if (index == _committed) return; // reveal fires once per question
        _committed = index;
        foreach (var s in _seats)
        {
            var a = _pending.FirstOrDefault(x => x.BotId == s.Bot.Id);
            var correct = a.ChoiceIndex == q.CorrectIndex;
            s.LastCorrect = a.ChoiceIndex is null ? false : correct;
            if (correct)
            {
                s.Streak += 1;
                s.Score += Scoring.Points(true, a.Seconds ?? budget, budget, s.Streak);
            }
            else s.Streak = 0;
        }
    }

    public IReadOnlyList<Seat> Standings => _seats.OrderByDescending(s => s.Score).ToList();
}
